### PR TITLE
Update phenotype import hive - e103

### PR DESCRIPTION
**Pull request by ima23**
fixes:
- gwas NR
- add seq_region_ids update for human
- AnimalQTL: import date only updated for imported species, species requiring remapping are skipped
- update meta_coord phenotype_feature related entries

address previous PR comments/other fixes:
- use hash for branch job flows
- define Human_var , Human_gene run_types
- move run_type from global to internal param 
- make finish module standalone

jira tickets: ENSVAR-3223, ENSVAR-3455, ENSVAR-3373, ENSVAR-2724, ENSVAR-2871, ENSVAR-3140

### CODE REVIEW

**Review by helensch — COMMENTED** [review]
Minor comments following initial review.
Review of pipeline flow being in second stage of review

File: modules/Bio/EnsEMBL/Variation/Pipeline/PhenotypeAnnotation/BasePhenotypeAnnotation.pm [hunk 1]
```diff
@@ -376,6 +376,36 @@ sub default_phenotype_class {
 #----------------------------
 # PUBLIC METHODS
 
+=head2 update_meta_coord
```
[the remark this answers is not in the document]
Comment (ima23) [reply modules/Bio/EnsEMBL/Variation/Pipeline/PhenotypeAnnotation/BasePhenotypeAnnotation.pm:379]: Yes, the method could be made more general and placed in a utils package. Let me know if you have suggestions of where, as it stands this is specific to this pipeline.

File: modules/Bio/EnsEMBL/Variation/Utils/SeqRegionUtils.pm [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  Arg[3]      : (optional) dry_run: only list the sql updates that would be applied; default 0
+  Example     :  use Bio::EnsEMBL::Variation::Pipeline::Utils::SeqRegionUtils qw(update_seq_region_ids);
+                update_seq_region_ids($core_dba, $variation_dba, 0)
+  Description : checks that seq_region name and seq_region_id are in sync
+                with the core db and applies any needed updates.
+  Exceptions  : None
+
+=cut
+
+sub update_seq_region_ids {
```
[the remark this answers is not in the document]
Comment (ima23) [reply modules/Bio/EnsEMBL/Variation/Utils/SeqRegionUtils.pm:61]: Yes, the module is outside the pipeline and anyone can use it for any species.

File: modules/Bio/EnsEMBL/Variation/Utils/SeqRegionUtils.pm [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+use strict;
+use warnings;
+
+use base qw(Exporter);
+
+our @EXPORT_OK = qw(update_seq_region_ids);
+
+
+=head2 update_seq_region_ids
```
[the remark this answers is not in the document]
Comment (ima23) [reply modules/Bio/EnsEMBL/Variation/Utils/SeqRegionUtils.pm:48]: Good point. Will create a ticket for it.

**Comment by ima23** [comment]
Thanks @helensch  for checking. I've made the minor comment changes and made a ticket re SeqRegionUtils for the raised points.

File: modules/Bio/EnsEMBL/Variation/Utils/SeqRegionUtils.pm (hunk 2 above)
[the remark this answers is not in the document]
Comment (helensch) [reply modules/Bio/EnsEMBL/Variation/Utils/SeqRegionUtils.pm:61]: I was querying if the data was available for all species to use this sub.

File: modules/Bio/EnsEMBL/Variation/Pipeline/PhenotypeAnnotation/CheckPhenotypeAnnotation.pm [hunk 4]
```diff
@@ -43,20 +43,29 @@ use File::Path qw(make_path);
...
+use Bio::EnsEMBL::Variation::Pipeline::PhenotypeAnnotation::Constants qw(SPECIES MOUSE MGI OMIA HUMAN HUMAN_VAR HUMAN_GENE CGC EGA ANIMALSET ANIMALQTL RGD ZFIN);
 
 my $source;
 my $workdir;
 my $species;
 my $report;
 my $count_ok = 1;
 
+my %groups_end_source = (
+  Human      => CGC,
```
Comment (helensch) [inline modules/Bio/EnsEMBL/Variation/Pipeline/PhenotypeAnnotation/CheckPhenotypeAnnotation.pm:55]: Should this be HUMAN ?

File: modules/Bio/EnsEMBL/Variation/Pipeline/PhenotypeAnnotation/Constants.pm [hunk 5]
```diff
@@ -54,37 +54,66 @@ our @EXPORT_OK = qw(
...
-  MOUSE     => 'Mouse',
+  MOUSE     => 'MOUSE',
 
   NONE      => 'NONE',
 };
 
+use constant GROUP_RUN_TYPES => (ANIMALSET => ['OMIA','AnimalQTL'],
+                                MOUSE     => ['IMPC', 'MGI'],
+                                HUMAN     => ['GWAS', 'EGA',
+                                              'Orhpanet', 'MIMmorbid',
```
Comment (helensch) [inline modules/Bio/EnsEMBL/Variation/Pipeline/PhenotypeAnnotation/Constants.pm:93]: Typo in "Orhpanet"

File: modules/Bio/EnsEMBL/Variation/Pipeline/PhenotypeAnnotation/ImportHuman.pm [hunk 6]
```diff
@@ -67,24 +90,11 @@ sub write_output {
...
-      $self->dataflow_output_id($self->param('output_ids'), 6);
-      $self->print_pipelogFH( "Passing to DDG2P import: ".scalar @{$self->param('output_ids')}." species\n") if $self->param('debug_mode');
-    } elsif ( $run_type eq CGC){
-      $self->dataflow_output_id($self->param('output_ids'), 7);
-      $self->print_pipelogFH( "Passing to CGC import: ".scalar @{$self->param('output_ids')}." species\n") if $self->param('debug_mode');
+    if ($source2branch{$run_type}){
+      $self->dataflow_output_id($self->param('output_ids'), $source2branch{$run_type});
+      $self->print_pipelogFH( "Passing to $run_type import: ".scalar @{$self->param('output_ids')}." species\n") if $self->param('debug_mode');
+    } else {
+      $self->print_pipelogFH("Runtype $run_type not supproted!\n");
```
Comment (helensch) [inline modules/Bio/EnsEMBL/Variation/Pipeline/PhenotypeAnnotation/ImportHuman.pm:97]: Typo in "supproted"
If run_type not supported,  should import continue?

**Review by helensch — COMMENTED** [review]
Review of flow changes in PR done. Minor comments.
In future would be useful to review/document:
-  pipeline configuration: hive_capacity=1, empty input_ids, empty flow_into
- use of autoflow branch in runnables
- dataflow/branches in CheckPhenotypeAnnotation

**Comment by ima23** [comment]
Thanks for the feedback @helens. PR is ready for review again. I have added a few more fixes based on the production run. I have addressed the minor comments and created two tickets relating to SeqRegionUtils, ENSVAR-3659, and eHive questions, ENSVAR-3666. Please feel free to add to the tickets anything that would be useful for future discussions.

File: modules/Bio/EnsEMBL/Variation/Pipeline/PhenotypeAnnotation/CheckPhenotypeAnnotation.pm [hunk 7]
```diff
@@ -105,58 +114,64 @@ sub run {
...
-          $self->dataflow_output_id($self->param('output_ids'), 2);
+      if ($run_type eq ANIMALSET && $source->{source_name} eq OMIA) {
+        my %animalQTL_species = map { $_ => 1 } @{$import_species{ANIMALQTL}};
+        # only forward jobs to AnimalQTL for AnimalQTL species
+        $fwd = 0 if (!defined($animalQTL_species{$species}));
+      }
+
+      if ($fwd) {
+        #generally another import_source module
+        $self->dataflow_output_id($self->param('output_ids'), 2);
```
[the remark this answers is not in the document]
Comment (helensch) [reply modules/Bio/EnsEMBL/Variation/Pipeline/PhenotypeAnnotation/CheckPhenotypeAnnotation.pm:156]: I now understand the module is controlling the flow to the different analysis. Suggest adding information on the what the branch flow is in the module.

File: modules/Bio/EnsEMBL/Variation/Pipeline/PhenotypeAnnotation/BasePhenotypeAnnotation.pm [hunk 8]
```diff
@@ -376,6 +376,36 @@ sub default_phenotype_class {
...
+
+sub update_meta_coord {
+  my $self = shift;
+
+  my $core_db_name = $self->core_db_adaptor->dbc()->dbname();
+  my $vdbh =  $self->variation_db_adaptor->dbc();
+  my $table_name ='phenotype_feature';
+
+  # delete previous phenotype_feature entries in update_meta_coord
+  my $mc_del_sth = $vdbh->do(qq[ DELETE FROM meta_coord WHERE table_name ='$table_name' ]);
```
Comment (helensch) [inline modules/Bio/EnsEMBL/Variation/Pipeline/PhenotypeAnnotation/BasePhenotypeAnnotation.pm:396]: The values returned by the do statements are not used. Variable suggests that it is a  statement handler.  do returns number of rows  or undef on error.

**Review by helensch — APPROVED** [review]
Thanks for the updates. I have added a ticket ENSVAR-3719 for suggestions for update_meta_coord.

**Comment by helensch** [comment]
merged into master and release/103

**Pull request merged**